internal/pkg/manifest/env.go

```diff
@@ -107,6 +108,32 @@ type environmentVPCConfig struct {
 	Subnets subnetsConfiguration `yaml:"subnets,omitempty"`
 }
 
+type securityGroupConfig struct {
+	Ingress []securityGroupRule `yaml:"ingress,omitempty"`
+	Egress  []securityGroupRule `yaml:"egress,omitempty"`
+}
+
+// IsEmpty returns true if there is no security group rule defined.
+func (cfg securityGroupConfig) IsEmpty() bool {
+	return len(cfg.Ingress) == 0 && len(cfg.Egress) == 0
+}
+
+type securityGroupRule struct {
+	CidrIP     string `yaml:"cidr"`
```

Comment:
This is sort of related to my other comment in `templates/environment/cf.yml`. I keep debating whether in general we will need to standardize these basic fields to have a pointer, so that we can distinguish an input of zero value from an absent input in env manifest, that is for example, to tell
```
- to_port: 0
  ip_protocol: "tcp"
```
from
```
- ip_protocol: "tcp"
```

If we use `*int`, then we would be able to; if we use `int` (like what you are doing now), then we won't be able to tell. 

For service manifest, we need to be able to tell them apart so that environment overrides are applied correctly; but this is not a thing in environment manifest. 

Up until now, it seems like for env manifest, it's not entirely necessary for us to distinguish zero-value vs. absent except for `bool` fields. But I don wonder though whether we should force every basic field to be a pointer field just to standardize it 💭

Reply:
That is a good call out actually. In the env manifest we do not have a consistent way of implementing this particularly thing. Currently, I have changed ToPort and FromPort to be pointers so that we can figure if customer has actually set this parameters in the manifest or not? And if not then set them as `-1` or `0`  whichever we decide in out other conversation [here](https://github.com/aws/copilot-cli/pull/3810#discussion_r931520948)